xorcery-dns-registration-azure/src/main/java/com/exoreaction/xorcery/service/dns/registration/azure/JsonApiClient.java

```diff
@@ -0,0 +1,198 @@
+package com.exoreaction.xorcery.service.dns.registration.azure;
+
+import com.exoreaction.xorcery.service.dns.registration.azure.model.AzureDnsRecordRequest;
+import com.exoreaction.xorcery.service.dns.registration.azure.model.AzureTokenResponse;
+import com.exoreaction.xorcery.service.jetty.client.CompletableFutureResponseListener;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.util.BytesRequestContent;
+import org.eclipse.jetty.client.util.FormRequestContent;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.util.Fields;
+
+import java.io.ByteArrayInputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+
+public class JsonApiClient {
+    private static final String APPLICATION_JSON = "application/json";
+    private static final String AZURE_OAUTH2_URL_PATTERN = "https://login.microsoftonline.com/%s/oauth2/v2.0/token";
+    private static final String AZURE_PRIVATE_DNS_URL_PATTERN = "https://management.azure.com/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/privateDnsZones/%s/%s/%s";
+    private static final String API_VERSION = "2018-09-01";
+    private final ObjectMapper objectMapper = new JsonMapper();
+    private final HttpClient httpClient;
+
+    public JsonApiClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    public CompletionStage<AzureTokenResponse> token(String azureTenantId, String clientId, String clientSecret) {
+        var future = new CompletableFuture<ContentResponse>();
+        var fields = new Fields();
+        fields.put("client_id", clientId);
+        fields.put("client_secret", clientSecret);
+        fields.put("scope", "https://management.azure.com/.default");
+        fields.put("grant_type", "client_credentials");
+        httpClient.newRequest(AZURE_OAUTH2_URL_PATTERN.formatted(azureTenantId))
+            .accept(APPLICATION_JSON)
+            .method(HttpMethod.POST)
+            .body(new FormRequestContent(fields))
+            .send(new CompletableFutureResponseListener(future));
+        return future.thenApply(cr -> {
+           try {
+               var json = objectMapper.readTree(new ByteArrayInputStream(cr.getContent()));
+               return new AzureTokenResponse((ObjectNode) json);
+           } catch (Throwable e) {
+               throw new CompletionException(e);
+           }
+        });
+    }
+
+    public CompletionStage<Void> putARecord(
```

Comment:
These are all repeating the same over and over, could take type (enum?) as an argument